src/app/ClusterData.h

```diff
@@ -0,0 +1,382 @@
+
+        /**
+         * Given a path handle to a node, a TLV writer, and booleans indicating whether the
+         * value is null or not present, get the data from the Cluster source that will build a
+         * report. If the path handle is not a leaf node, IM will handle writing values to
+         * the writer (like opening containers etc). If a non-leaf
+         * node is null or not present, IM will not call GetData for its children.
+         *
+         * This function will only be called for handles that are leafs. The expectation is that any Clusters with handles that
+         * have those options enabled will implement appropriate logic to populate
+         * aIsPresent.
+         *
+         * @param[in] aHandle       The AttributePathHandle in question.
+         *
+         * @param[in] aTagToWrite   The tag to write for the aHandle.
+         *
+         * @param[inout] aWriter    The writer to write TLV elements to.
+         *
+         * @param[out] aIsPresent   Is aHandle present? If no and if aHandle
+         *                          is not a leaf, IM will skip over the
+         *                          path and its children.
+         *
+         *
+         * @retval #CHIP_NO_ERROR On success.
+         * @retval other           Was unable to retrieve data and write it into the writer.
+         */
+        virtual CHIP_ERROR GetData(AttributePathHandle aHandle, uint64_t aTagToWrite, chip::TLV::TLVWriter & aWriter,
+                                   bool & aIsPresent) = 0;
+        virtual ~IGetDataDelegate()                   = default;
+    };
+
+    /**
+     * Given a path handle and a reader positioned on the corresponding attribute data element, process the data buffer pointed to
+     * by the reader and store it into the sink by invoking the SetLeafData call whenever a leaf data item is encountered.
+     *
+     * @retval #CHIP_NO_ERROR On success.
+     * @retval other           Encountered errors parsing/processing the data.
+     */
+    CHIP_ERROR StoreData(AttributePathHandle aHandle, chip::TLV::TLVReader & aReader, ISetDataDelegate * apSetDataDelegate) const;
+
+    /**
+     * Given a path handle and a writer position on the corresponding attribute data element, retrieve leaf data from the source and
+     * write it into the buffer pointed to by the writer in a schema compliant manner.
+     *
+     * @retval #CHIP_NO_ERROR On success.
+     * @retval other           Encountered errors writing out the data.
+     */
+    CHIP_ERROR RetrieveData(AttributePathHandle aHandle, uint64_t aTagToWrite, chip::TLV::TLVWriter & aWriter,
+                            IGetDataDelegate * apGetDataDelegate) const;
+    /**********
+     *
+     * Schema Query Functions
+     *
+     *********/
+
+    /**
+     * Returns true if the handle refers to a leaf node in the schema tree.
+     *
+     * @retval bool
```

Comment:
please clean up all the comments. the `@retval` comments here and below just use up vertical space.